man/mlgsc.1.md
% MLGSC(1) MLgsc User Manuals
% Thomas Junier
% December 16, 2015

# NAME

mlgsc - maximum -likelihood general sequence classifier

# SYNOPSIS

mlgsc -h

mlgsc [*options*] <*queries filename*> <*model*> 

# DESCRIPTION

`mlgsc` classifies sequences according to a model produced by `mlgsc_train(1)`.
The queries file must be in Fasta format. Of course, they are expected to be
homologous to the reference sequences used in training the model.


# OPTIONS

-A, \--no-align
:   Do not align the input sequences. If you have input sequences that are
    already aligned (to the model, or from the same multiple alignment that the
    model was trained with) This can save substantial run time as the alignment
    step is often the bottleneck (especially with small numbers of target taxa),
    since it is performed by full dynamic programming.
    The default is to align.

-e, \--ER-cutoff *NUM*
:   Stop classifying a particular query if a node has an ER smaller than *NUM*.
    Intuitively, this means that `mlgsc` has reached a node at which cannot
    decide which of its children is best, and therefore stops. 
    The default is not to apply any cutoff (i.e., a cutoff of 0).

-f, \--output-format *STRING*
:   A printf(1)-like format string that specifies the format of the output
    elements. The recognized placeholders are:
   
    %%: literal '%'
    
    %a: the aligned query sequence (useful fo troubleshooting)

    %l: the length of the (unaligned) query sequence
    
    %h: the Fasta headerg
    
    %i: the query sequence's ID, taken to be the first word in the header line
    
    %m: the lowest evidence ratio found for the current query
    
    %P: the predicted taxon 
    
    %p: the whole path through the phylogeny, from the root to the predicted
    taxon. Path elements can be formatted via option `-s`.
    
    %q: the query sequence, unaligned
    
    %s: the score of the query against the predicted taxon
    
    The default is "%h -> %p", that is, the query header followed by the whole
    path. To see just the ID and the predicted taxon, for example, use `-f "%i
    %P"`.
    
-h, \--help
:   print a short help message, and exit successfully.

-M, --mask-mode *n|t*
:   How to mask the query sequence (if at all). This causes some portions of the
    query sequence to be ignored when scoring against the model (at all nodes in
    the tree). The possible values are:

    n (none): do not mask - use the whole sequence for scoring.
    
    t (trim): ignore the leading and trailing gapped regions. This is useful
    when the query sequence is substantially shorter than the model, yet has a
    good local match to it. This will divide the sequence into a matching region
    flanked by to regions which are mostly gaps. The boundaries between the
    regions are determined by a hidden Markov model, and those gapped regions
    will be effectively trimmed from the query sequence.
    
    The default is not to mask.

-m, --traversal-mode *b|a|INT*
:   Determines the search strategy, that is, how the model tree is traversed.
    Possible values are:

    b (best): at each inner node, choose the child with the _best_ score. This
    is the fastest option, but offers no error recovery.
    
    a (all): show _all_ paths through the tree. This is useful when
    troubleshooting, as it helps determine at which point the classifier makes a
    wrong choice. The nodes actually followed always shows an evidence ratio of
    zero.
    
    INT: the number supplied is used as an evidence ratio (ER) cutoff. Instead
    of searching only the child with the best score, as in "best" mode, search
    all children with an ER (with respect to the best-scoring child) not greater
    than the threshold (and take the best-scoring taxon among all these). This
    allows some error recovery, as it can happen that the correct taxon lies on
    a branch that is not attached to the best-scoring child. Of course,
    exploring more branches takes more time.

-s, --step-format
:   Like `-f`, but specifies how each element of the path (from the root to the
    target taxon) is formatted. Valid placeholders are:
    
    '%%': literal '%'
    
    '%t': taxon name (= tree node label)
    
    '%s': evidence ratio
    
    '%b': score of best-scoring child of the current node

    The default format is "%t (%s)", that is, the taxon name followed by the
    evidence ratio in parentheses.

# EXAMPLES

Data for all examples can be found in the _data_ directory of the MLgsc
distribution.

## Example 1

```
$ mlgsc Spo0A.pep Spo0A.bcls
```

This classifies the sequences in file Spo0A.pep according to the classifier in
file Spo0A.bcls. The output looks like this:

```
ID_001 Bacillus -> Bacilli (56); Bacillaceae (97); Bacillus (87)
ID_002 Clostridium -> Clostridia (106); Clostridiales (138); Clostridium (83)
ID_003 Clostridium -> Clostridia (106); Clostridiales (138); Clostridium (83)
ID_004 Bacillus -> Bacilli (79); Bacillaceae (95); Bacillus (99)
ID_005 Bacillus -> Bacilli (84); Bacillaceae (89); Bacillus (104)
ID_006 Bacillus -> Bacilli (84); Bacillaceae (89); Bacillus (104)
```

## Example 2

```
$ mlgsc -f "%i,%P,%s" Spo0A.pep  Spo0A.bcls
```

As in Example 1, but using option **-f** to change the output format. In this
case, we request the sequence id (**%i**), the predicted reference taxon
(**%P**), and the score (**%s**), separated by commas (that is, in CSV).

# SEE ALSO

`mlgsc_train` (1), `mlgsc_xval` (1) and `mlgsc_dump` (1).

MLgsc is available form  <https://github.com/tjunier/mlgsc>

A technical description of how __mlgsc__ classifies queries is found in the
following article:

Junier T, Hervé V, Wunderlin T, and Junier P (2015) MLgsc: a
Maximum-Likelihood, general sequence classifier. PLoS One <http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0129384>

Please cite this paper if you use MLgsc in published research. Thanks!




src/mlgsc.hs
-- General Sequence Classifier

module Main (main) where

-- Classifies a set of sequences using a classifier produced by gsc_mk.  
-- module Main where

import System.IO (hPutStrLn, stderr)
import System.Environment (getArgs)
import qualified Data.Text.Lazy as LT
import qualified Data.Text.Lazy.IO as LTIO
import qualified Data.Text as ST
import qualified Data.Text.IO as STIO
import Text.Printf
import qualified Text.Parse as TP
import Control.Applicative
import Control.Monad.Reader
import Options.Applicative

import Data.Maybe
import Data.Map (Map, fromList, findWithDefault)
import Data.Binary (decodeFile)
import Data.Tree
import Data.Char
import Data.List
import Data.Ord

import MlgscTypes
import FastA
import Trim
import PWMModel
import Align
import Classifier (Classifier(..), classifySequence,
    classifySequenceMulti, classifySequenceAll,
    StoredClassifier(..))
import Output

data TreeTraversalMode = BestTraversal
                       | FullTraversal
                       | RecoverTraversal Int

data MaskMode = None | Trim

data SingleNodeScope =  NodesChildren | WholeSubtree
    
data Params = Params {
                optTreeTraversalMode    :: TreeTraversalMode
                , optNoAlign            :: Bool
                , optOutFmtString       :: String
                , optStepFmtString      :: String
                , optERCutoff           :: Int   -- for Best mode (TODO: could be an argument to the BestTraversal c'tor)
                , optMaskMode           :: MaskMode
                , opt1NodeScope         :: SingleNodeScope
                , posParams             :: [String]
                }

parseTreeTraversal :: Monad m => String -> m TreeTraversalMode
parseTreeTraversal optString
    | 'b' == initC = return BestTraversal
    | 'a' == initC = return FullTraversal
    | otherwise    = do
                        let (Right num,_) = TP.runParser TP.parseDec optString
                        -- TODO: handle bad parse
                        return $ RecoverTraversal num
    where initC = toLower $ head optString

parseMaskMode :: Monad m => String -> m MaskMode
parseMaskMode optString
    | 't' == initC = return Trim
    | 'n' == initC = return None
    | otherwise = return None -- TODO: warn about unrecognized opt
    where initC = toLower $ head optString

parseSingleNodeScope :: Monad m => String -> m SingleNodeScope
parseSingleNodeScope optString
    | 'c' == initC = return NodesChildren
    | 'w' == initC = return WholeSubtree
    -- TODO: handle others
    where initC = toLower $ head optString

parseOptions :: Parser Params
parseOptions = Params
                <$> option (str >>= parseTreeTraversal)
                    (long "traversal-mode"
                    <> short 'm'
                    <> help "tree traversal mode (b|a|<int>)"
                    <> value BestTraversal)
                <*> switch
                    (long "no-align"
                    <> short 'A'
                    <> help "do not align query sequences")
                <*> option str
                    (long "output-format"
                    <> short 'f'
                    <> help "printf-like format string for output."
                    <> value "%h -> %p")
                <*> option str
                    (long "step-format"
                    <> short 's'
                    <> help "printf-like format string for step (path element)"
                    <> value "%t (%s)")
                <*> option auto
                    (long "ER-cutoff"
                    <> short 'e'
                    <> help "drop clades with ER lower than this"
                    <> value 0)
                <*> option (str >>= parseMaskMode)
                    (long "mask-mode"
                    <> short 'M'
                    <> help "mask aligned query: n)one* | t)trim"
                    <> value None)
                <*> option (str >>= parseSingleNodeScope)
                    (long "single-node-scope"
                    <> short 'S'
                    <> help "only score vs s) single node or its c)hildren"
                    <> value NodesChildren)
                <*> many (argument str (metavar "<query seq file> <classifier filename> [clade name]"))

parseOptionsInfo :: ParserInfo Params
parseOptionsInfo = info (helper <*> parseOptions) 
                    ( fullDesc
                    <> progDesc "classify sequences according to a model"
                    <> Options.Applicative.header
                        "mlgsc - maximum-likelihood general sequence classifier")


-- Some common format options have names (e.g.  "simple" -> "%h -> %P").
-- These must be translated using the following map (some short forms are
-- possible)

fmtMap :: Map String String
fmtMap = fromList [
                    ("minimal", "%i\t%P"),
                    ("min", "%i\t%P"),
                    ("m", "%i\t%P"),
                    ("simple", "%h -> %P (%s)"),
                    ("s", "%h -> %P (%s)")
                ]

translateFmtKw :: Params -> IO Params
translateFmtKw params = do
    let origFmt = optOutFmtString params
    let fmt = findWithDefault origFmt origFmt fmtMap
    return params { optOutFmtString = fmt }
                

main :: IO ()
main = do
    params <- execParser parseOptionsInfo >>= translateFmtKw
    let positionalParams = posParams params
    let (queryFname, clsfrFname, singleNode) =
            case positionalParams of
                [q,c,n] -> (q,c,Just n)
                [q,c]   -> (q,c,Nothing)
    queryFastA <- LTIO.readFile queryFname
    let queryRecs = fastATextToRecords queryFastA
    storedClassifier <- (decodeFile clsfrFname) :: IO StoredClassifier
    let (StoredClassifier
            origClassifier@(PWMClassifier modTree scale)
            _) = storedClassifier
    let rootMod = rootLabel modTree
    -- Some options reduce the classifier to a subclade
    let classifier =
            case singleNode of
                 Nothing -> origClassifier
                 Just node -> case subClsf of
                                  Nothing -> error "node not found"
                                  (Just c) -> c
                        where subClsf = subsetClassifier origClassifier (ST.pack node) (opt1NodeScope params) 
    -- TODO: replace the magic "-2" below by a meaningful
    -- constant/param
    let scoringScheme =
            ScoringScheme (-2) (scoringSchemeMap
            (absentResScore rootMod))
    let headers = map FastA.header queryRecs
    let processedQueries =
            map (
                -- All the transformations from Fasta record to
                -- ready-to-score query. Note that some (mb*)
                -- depend on params among other things.
                mbMask params
                . mbAlign params scoringScheme rootMod
                . ST.toUpper
                . LT.toStrict
                . FastA.sequence
            ) queryRecs
    let outlines =
            case optTreeTraversalMode params of
                BestTraversal ->
                    bestTraversal params classifier queryRecs
                    processedQueries
                (RecoverTraversal _) ->
                    recoverTraversal params classifier queryRecs
                    processedQueries
                FullTraversal -> fullTraversal params classifier
                    queryRecs processedQueries

    mapM_ STIO.putStrLn outlines

subsetClassifier :: Classifier -> CladeName -> SingleNodeScope ->
    Maybe Classifier
subsetClassifier (PWMClassifier modtree scale) name scope = 
            case trimmedSubtree of
                (Just s) -> Just $ PWMClassifier s scale
                Nothing  -> Nothing
            where   trimmedSubtree = case scope of
                                        NodesChildren -> subtree >>= trim
                                        WholeSubtree -> subtree
                    subtree = subTreeNamed name modtree

subTreeNamed :: CladeName -> (Tree PWMModel) -> Maybe (Tree PWMModel)
subTreeNamed name subtree@(Node model kids)
    | name == cladeName model = Just subtree
    | otherwise =
        case kids of
           [] -> Nothing
           (k:ks) -> listToMaybe $ catMaybes $ map (subTreeNamed name) kids

trim:: Tree a -> Maybe (Tree a)
trim leaf@(Node _ []) = Just leaf
trim (Node x kids) = Just $ Node x $ catMaybes $ map toLeaf kids

toLeaf :: Tree a -> Maybe (Tree a)
toLeaf n = Just $ n {subForest = []}

-- Returns an alignment step (technically, a ST.Text -> ST.Text
-- function), or just id if option 'optNoAlign' is set. The parens in the
-- signature are not strictly necessary, but emphasize the fact that we
-- return a function.

mbAlign :: Params -> ScoringScheme -> PWMModel -> (Sequence -> Sequence)
mbAlign params scsc rootMod =
    if optNoAlign params
            then id
            else msalign scsc rootMod

-- Returns a masking step (a ST.Text -> ST.Text function),
-- or just id if no masking was requested.

mbMask :: Params -> (ST.Text -> ST.Text)
mbMask params =
    case optMaskMode params of
        None -> id
        Trim -> trimSeq

bestTraversal :: Params -> Classifier -> [FastA] -> [Sequence] -> [ST.Text]
bestTraversal params classifier queryRecs processedQueries =
    getZipList $ (formatResultWrapper params)
        <$> ZipList queryRecs
        <*> ZipList processedQueries
        <*> ZipList predictions 
    where
        log10ER = optERCutoff params
        predictions = map (classifySequence classifier log10ER) processedQueries

recoverTraversal :: Params -> Classifier -> [FastA] -> [Sequence] -> [ST.Text]
recoverTraversal params classifier queryRecs processedQueries =
    getZipList $ (formatResultWrapper params)
        <$> ZipList queryRecs
        <*> ZipList processedQueries
        <*> ZipList bestPredictions
    where
        (RecoverTraversal tieThreshold)  = optTreeTraversalMode params
        bestPredictions = map getBestPrediction predictions
        predictions = map (classifySequenceMulti classifier tieThreshold) processedQueries
    
getBestPrediction :: [Trail] -> Trail
getBestPrediction trails = maximumBy (comparing lastScore) trails

lastScore :: Trail -> Score
lastScore trail = bestScore $ last trail 

fullTraversal :: Params -> Classifier -> [FastA] -> [Sequence] -> [ST.Text]
fullTraversal params classifier queryRecs processedQueries =
    concat $ getZipList $ (fullTraversalFmt1Query params)
        <$> ZipList queryRecs
        <*> ZipList processedQueries
        <*> ZipList predictions
    where
        predictions = map (classifySequenceAll classifier) processedQueries
    
fullTraversalFmt1Query :: Params -> FastA -> Sequence -> [Trail] -> [ST.Text]
fullTraversalFmt1Query params queryRec processQuery trails =
    map (formatResultWrapper params queryRec processQuery) trails

{- I like to apply the output formatter in aplicative style to the lists of
- arguments. However, I'm not sure how to make this play with the Reader
- monad, except by using a wrapper like below. -}

formatResultWrapper :: Params -> FastA -> Sequence -> Trail -> ST.Text  
formatResultWrapper params query alnQry trail =
    runReader (formatResultReader query alnQry trail) params

{- This works, but it's not quite clear what should go in Output,
 - OutputFormatStringParser, or just plain here in Main. -}

formatResultReader :: FastA -> Sequence -> Trail -> Reader Params ST.Text 
formatResultReader query alnQry trail = do
    fmtString    <- asks optOutFmtString
    stepFmtString <- asks optStepFmtString
    let (Right format) = parseOutputFormatString fmtString 
    let (Right stepfmt) = parseStepFormatString stepFmtString
    return $ ST.concat $ map (evalFmtComponent query alnQry trail stepfmt) format

